Some editorial changes + from_prior now MUST for DID Rotation + made key words uppercase (#260)
decentralized-identity/didcomm-messaging: docs/spec-files/message_structure.md

When higher-level protocols are built atop [DIDComm Messaging](.), applications remove the protective envelope(s) and process the plaintext that's inside. Specifications for higher-level protocols typically document message structure and provide examples in this format; protective envelopes are assumed but ignored as a low-level detail.

The media type for a generic **DIDComm plaintext message** MUST be `application/didcomm-plain+json`. **DIDComm plaintext messages** are also correctly understood as [JWM](https://tools.ietf.org/html/draft-looker-jwm-01) content (see [Plaintext Message Structure](#plaintext-message-structure), below) and the media type MUST be set in the `typ` [property](https://tools.ietf.org/html/rfc7515#section-4.1.9) of the **DIDComm plaintext message** header.

> When persisted as a file or attached as a payload in other contexts, the file extension for **DIDComm plaintext messages** SHOULD be `dcpm`, giving a globbing pattern of `*.dcpm`; this SHOULD be be read as "Star Dot D C P M" or as "D C P M" files. We imagine people will reference this media type by saying, "I am looking at a DIDComm Plaintext Message file", or "This database record is in DIDComm Plaintext format", or "Does my editor have a DIDComm Plaintext Message plugin?" A possible icon for this file format depicts green JSON text in a message bubble ([svg](../collateral/dcpm.svg) | [256x256](../collateral/dcpm-256.png) | [128x128](../collateral/dcpm-128.png) | [64x64](../collateral/dcpm-64.png)):

Reviewer: ```suggestion
When persisted as a file or attached as a payload in other contexts, the file extension for **DIDComm plaintext messages** MUST be `dcpm`, giving a globbing pattern of `*.dcpm`; this MUST be be read as "Star Dot D C P M" or as "D C P M" files. We imagine people will reference this media type by saying, "I am looking at a DIDComm Plaintext Message file", or "This database record is in DIDComm Plaintext format", or "Does my editor have a DIDComm Plaintext Message plugin?" A possible icon for this file format depicts green JSON text in a message bubble ([svg](../collateral/dcpm.svg) | [256x256](../collateral/dcpm-256.png) | [128x128](../collateral/dcpm-128.png) | [64x64](../collateral/dcpm-64.png)):
```

Any reason this shouldn't be a MUST here? Happy to move this to a separate issue and discussion as well since it's not an editorial change in nature. Second thing I thought about when reviewing this: Do we need to register this file type with IANA?

Author: the goal of this PR was just purely editorial. imo, we should review all normative statements anyways and if changes are required, we should have separate PRs.

Author: Resolving this discussion since we have a separate ticket for file extensions #279

When a message is *both* signed and encrypted, the plaintext is signed, and then the signed envelope is encrypted. The opposite order is not used, since it would imply that the signer committed to opaque data (which is unsafe and undermines non-repudiation).

The [media type](https://tools.ietf.org/html/rfc6838) of a **DIDComm signed message** MUST be `application/didcomm-signed+json`.

The media type of the envelope MUST be set in the `typ` [property](https://tools.ietf.org/html/rfc7515#section-4.1.9) of the JWS.

> When persisted as a file or attached as a payload in other contexts, the file extension for **DIDComm signed messages** SHOULD be `dcsm`, giving a globbing pattern of `*.dcsm`; this SHOULD be be read as "Star Dot D C S M" or as "D C S M" files. A possible icon for this media type depicts a signed envelope ([svg](../collateral/dcsm.svg) | [256x256](../collateral/dcsm-256.png) | [128x128](../collateral/dcsm-128.png) | [64x64](../collateral/dcsm-64.png)):

Reviewer: ```suggestion
When persisted as a file or attached as a payload in other contexts, the file extension for **DIDComm signed messages** MUST be `dcsm`, giving a globbing pattern of `*.dcsm`; this MUST be be read as "Star Dot D C S M" or as "D C S M" files. A possible icon for this media type depicts a signed envelope ([svg](../collateral/dcsm.svg) | [256x256](../collateral/dcsm-256.png) | [128x128](../collateral/dcsm-128.png) | [64x64](../collateral/dcsm-64.png)):
```

Same question about registration here

Author: the goal of this PR was just purely editorial. imo, we should review all normative statements anyways and if changes are required, we should have separate PRs.

Author: Resolving this discussion since we have a separate ticket for file extensions #279

![DIDComm Signed Message Icon](../collateral/dcsm-128.png)

### DIDComm Encrypted Message

A **DIDComm encrypted message** is an encrypted [JWM (JSON Web Messages)](https://tools.ietf.org/html/draft-looker-jwm-01) and hides its content from all but authorized recipients, discloses and proves the sender to exactly and only those recipients, and provides integrity guarantees. It is important in privacy-preserving routing. It is what normally moves over network transports in [DIDComm Messaging](.) applications, and is the safest format for storing [DIDComm Messaging](.) data at rest.

The [media type](https://tools.ietf.org/html/rfc6838) of a non-nested **DIDComm encrypted message** MUST be `application/didcomm-encrypted+json`.

> > Note: If future versions of this spec allow binary encodings, variations like `application/didcomm-encrypted+cbor` (see [CBOR RFC 7049, section 7.5](https://tools.ietf.org/html/rfc7049#section-7.5)), `application/didcomm-encrypted+msgpack`, or `application/didcomm-encrypted+protobuf` may become reasonable. In the future, specifications that encompass comm patterns other than messaging &mdash; DIDComm Multicast or DIDComm Streaming, for example &mdash; might use a suffix: `application/didcomm-encrypted-multicast` or similar.

Reviewer: ```suggestion
> Note: If future versions of this spec allow binary encodings, variations like `application/didcomm-encrypted+cbor` (see [CBOR RFC 7049, section 7.5](https://tools.ietf.org/html/rfc7049#section-7.5)), `application/didcomm-encrypted+msgpack`, or `application/didcomm-encrypted+protobuf` may become reasonable. In the future, specifications that encompass communications patterns other than messaging &mdash; DIDComm Multicast or DIDComm Streaming, for example &mdash; might use a suffix: `application/didcomm-encrypted-multicast` or similar.
```

The media type of the envelope MUST be set in the `typ` [property](https://tools.ietf.org/html/rfc7516#section-4.1.11) of the JWE.

> When persisted as a file or attached as a payload in other contexts, the file extension for **DIDComm encrypted messages** SHOULD be `dcem`, giving a globbing pattern of `*.dcem`; this SHOULD be read as "Star Dot D C E M" or as "D C E M" files. A possible icon for this file format depicts an envelope with binary overlay, protected by a lock ([svg](../collateral/dcem.svg) | [256x256](../collateral/dcem-256.png) | [128x128](../collateral/dcem-128.png) | [64x64](../collateral/dcem-64.png)):

Reviewer: ```suggestion
When persisted as a file or attached as a payload in other contexts, the file extension for **DIDComm encrypted messages** MUST be `dcem`, giving a globbing pattern of `*.dcem`; this MUST be read as "Star Dot D C E M" or as "D C E M" files. A possible icon for this file format depicts an envelope with binary overlay, protected by a lock ([svg](../collateral/dcem.svg) | [256x256](../collateral/dcem-256.png) | [128x128](../collateral/dcem-128.png) | [64x64](../collateral/dcem-64.png)):
```

Same question about registration

Author: the goal of this PR was just purely editorial. imo, we should review all normative statements anyways and if changes are required, we should have separate PRs.

Author: Resolving this discussion since we have a separate ticket for file extensions #279

### Message Headers

A **DIDComm plaintext message** conveys most of its application-level data inside a JSON `body` object that is a direct child of the message root. The structure inside `body` is predicted by the value of the message's `type` attribute, and varies according to the definition of the protocol-specific message in question. Each `type` of message will have its own `body`. 

However, some attributes are common to many different message types. When metadata about a message means the same thing regardless of context, and when it is susceptible to generic rather than message-specific handling, that metadata can be placed in **headers**. Headers are siblings of `body` and may be added to any message type. They are encrypted and decrypted along with `body` and therefore have an identical audience.

Headers in [DIDComm Messaging](.) are intended to be extensible in much the same way that headers in HTTP or SMTP are extensible. A few headers are predefined:

- **id** - REQUIRED. Message ID. The `id` attribute value MUST be unique to the sender.

> - **type** - REQUIRED. Plaintext message type, useful for message handling in application-level protocols. The `type` attribute value MUST be a valid [Message Type URI](protocols.md#message-type-uri), that when resolved gives human readable information about the message. The attribute's value predicts the content in the `body` of the message, for example the presence of other attributes and how they SHOULD be processed.

Reviewer: ```suggestion
- **type** - REQUIRED. Plaintext message type, useful for message handling in application-level protocols. The `type` attribute value MUST be a valid [Message Type URI](protocols.md#message-type-uri), that when resolved gives human readable information about the message. The attribute's value SHOULD predict the content in the `body` of the message.
```

> - **typ** - OPTIONAL. Media type of the JWM content. MUST be set to `application/didcomm-plain+json` for **DIDComm plaintext messages**. This tells generic JOSE libraries that the [JWM](https://tools.ietf.org/html/draft-looker-jwm-01) in question is [DIDComm Messaging](.) rather than a different [JWM](https://tools.ietf.org/html/draft-looker-jwm-01) type and is therefore helpful if messages will be processed in a system that supports a variety of [JWMs](https://tools.ietf.org/html/draft-looker-jwm-01).

Reviewer: This normative MUST statement appears redundant to the one listed on line 17 above in this file. Do we need both?

Author: > This normative MUST statement appears redundant to the one listed on line 17 above in this file. Do we need both?

Good find, I agree.

However, some attributes are common to many different message types. When metadata about a message means the same thing regardless of context, and when it is susceptible to generic rather than message-specific handling, that metadata can be placed in **headers**. Headers are siblings of `body` and may be added to any message type. They are encrypted and decrypted along with `body` and therefore have an identical audience.

Headers in [DIDComm Messaging](.) are intended to be extensible in much the same way that headers in HTTP or SMTP are extensible. A few headers are predefined:

- **id** - REQUIRED. Message ID. The `id` attribute value MUST be unique to the sender.

- **type** - REQUIRED. Plaintext message type, useful for message handling in application-level protocols. The `type` attribute value MUST be a valid [Message Type URI](protocols.md#message-type-uri), that when resolved gives human readable information about the message. The attribute's value SHOULD predict the content in the `body` of the message.

> - **typ** - OPTIONAL. Media type of the JWM content.

Author: @kdenhartog could you pls check whether you are ok with the current language. I removed redundant normative statement on `typ` MUST be `application/didcomm-plain+json`.

    The following table provides an overview of using the media type properties in supported [DIDComm Messaging](.) messages:

    | DIDComm message | `typ` |
    |-----------------|-------|
    | Authcrypted and/or anoncrypted | `application/didcomm-encrypted+json`|
    | Signed and anoncrypted | `application/didcomm-encrypted+json`|
    | Signed | `application/didcomm-signed+json` |

>     | Plaintext | `application/didcomm-plain+json`|

Author: @kdenhartog could you pls check if you are ok with those changes. Idea is to make the table more readable (complete).

- **expires_time** - OPTIONAL. Message Expired Time. The `expires_time` attribute is used for the sender to express when they consider the message to be expired, expressed in UTC Epoch Seconds (seconds since 1970-01-01T00:00:00Z UTC) [link](1970-01-01T00:00:00Z UTC). This attribute signals when the message is considered no longer valid by the sender. When omitted, the message is considered to have no expiration by the sender.

With respect to headers, [DIDComm Messaging](.) follows the extensibility pattern established by the JW* family of standards. A modest inventory of predefined "header" fields is specified, as shown above. Additional fields with unreserved names can be added at the discretion of producers and consumers of messages; any software that doesn't understand such fields SHOULD ignore them and MUST NOT fail because of their inclusion in a message. This is appropriate for a simple, flat data model.

> Aligning with [RFC 6648](https://tools.ietf.org/html/rfc6648.html), [DIDComm Messaging](.) explicitly rejects the `X-*` headers convention that creates divergent pseudo-standards; if a new header needs broad support, proper standardization is required. Alternatively, a JSON-LD `@context` header can be added, providing namespacing for fields other than those predefined in the specification. Since we expect header fields to be small in number and modest in complexity, we expect this sort of powerful extensibility to be unnecessary in most cases.

Reviewer: What is the proper method for standardizing this? I think it's wise for us to not use "must" here since it's not going to be enforceable with code, but unsure what this implies now.

Reviewer: ```suggestion
Aligning with [RFC 6648](https://tools.ietf.org/html/rfc6648.html), [DIDComm Messaging](.) explicitly rejects the `X-*` headers convention that creates divergent pseudo-standards; if a new header needs broad support, proper standardization is required. Since we expect header fields to be small in number and modest in complexity, we expect this sort of powerful extensibility to be unnecessary in most cases.
```

I've worked with JSON-LD enough that I think this is opening a can of worms that's unnecessary. I'm thinking we strike these mentions here or double down on JSON-LD and specify how it works. Given JWMs are designed to be simple JSON my preference is to strike mentions of it.

Author: > What is the proper method for standardizing this? I think it's wise for us to not use "must" here since it's not going to be enforceable with code, but unsure what this implies now.

Yep, that is why I avoided the MUST and I apologise that I didn't put this in the description of the PR.

When a [DID](https://www.w3.org/TR/did-core/) is rotated, the new [DID](https://www.w3.org/TR/did-core/) is put into immediate use encrypting the message, and one additional attribute MUST be included as a message header:

- **from_prior**: REQUIRED. A JWT, with with `sub`: new [DID](https://www.w3.org/TR/did-core/) and `iss`: prior [DID](https://www.w3.org/TR/did-core/), with a signature from a key authorized by prior [DID](https://www.w3.org/TR/did-core/).

> When a message is received from an unknown [DID](https://www.w3.org/TR/did-core/), the recipient SHOULD check for existence of the `from_prior` header. The JWT in the`from_prior` attribute is used to extract the prior [DID](https://www.w3.org/TR/did-core/) (`iss`) and is checked to verify the validity of the rotation. The recipient then associates the message with context related to the known sender. The new [DID](https://www.w3.org/TR/did-core/) and associated [DID](https://www.w3.org/TR/did-core/) Document information SHOULD be used for further communication.

Reviewer: I'm thinking we should be upgrading these SHOULD to MUST. I can't find any reason that it shouldn't be considered and I think that ignoring it could introduce some weird security issues.

Author: I fully agree but the goal of this PR was just purely editorial. imo, we should review all normative statements anyways and if changes are required, we should have separate PRs.

Author: I'll create a separate ticket: #278

Author: Resolving this since we have now a separate ticket.

The validity of the [DID Rotation](#did-rotation) is verified by checking the JWT signature against the key indicated in the `kid` header parameter. The indicated key MUST be authorized in the [DID Document](https://www.w3.org/TR/did-core/#dfn-did-documents) of the prior [DID](https://www.w3.org/TR/did-core/) (`iss`).

> The `from_prior` attribute SHOULD be included in messages sent until the party rotating receives a message sent to the new [DID](https://www.w3.org/TR/did-core/). If multiple messages are received to containing the rotation headers after being processed by the recipient, they may be ignored.

Reviewer: ```suggestion
The `from_prior` attribute MUST be included in messages sent until the party rotating receives a message sent to the new [DID](https://www.w3.org/TR/did-core/). If multiple messages are received to containing the rotation headers after being processed by the recipient, they MAY be ignored.
```

Author: fully agree

#### Rotation Limitations

> - This rotation method does not cover cases where a multi-sig is required. Rotations with such requirements SHOULD use a more expressive protocol.

Reviewer: ```suggestion
- This rotation method only supports the case where a new [DID](https://www.w3.org/TR/did-core/) is used, replacing an old [DID](https://www.w3.org/TR/did-core/) which is no longer used. Adjustments to [DIDs](https://www.w3.org/TR/did-core/) used between different parties that does not fit this narrow use are expected to define a separate protocol to do so.
```
